### DISCUSSION TITLE
Does SWIG support passing an `std::unique_ptr` parameter "by-value"?

### DISCUSSION THREAD

**Discussion by N-Dekker** [swig/swig#2854] [Q&A] [answered]
I'm glad to see that SWIG now supports `std::unique_ptr`. ITK has very recently upgraded to using a very recent version from the "master" branch of SWIG, from 2024-03-26 (pull request https://github.com/InsightSoftwareConsortium/ITK/pull/4541 by @thewtex). Unfortunately it appears that a use case that I originally proposed at https://github.com/InsightSoftwareConsortium/ITK/pull/4533, introducing a member function `OverrideBoundaryCondition(std::unique_ptr<BoundaryConditionType>)` still got compile errors, like:

```
/home/.../Wrapping/Modules/ITKImageGradient/itkGradientImageFilterPython.cpp: In instantiation of 'SwigValueWrapper<T>& SwigValueWrapper<T>::operator=(const T&) [with T = std::unique_ptr<itk::ImageBoundaryCondition<itk::Image<float, 2>, itk::Image<float, 2> >, std::default_delete<itk::ImageBoundaryCondition<itk::Image<float, 2>, itk::Image<float, 2> > > >]':
/home/.../Wrapping/Modules/ITKImageGradient/itkGradientImageFilterPython.cpp:4587:15:   required from here
/home/.../Wrapping/Modules/ITKImageGradient/itkGradientImageFilterPython.cpp:3856:66: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = itk::ImageBoundaryCondition<itk::Image<float, 2>, itk::Image<float, 2> >; _Dp = std::default_delete<itk::ImageBoundaryCondition<itk::Image<float, 2>, itk::Image<float, 2> > >]'
 3856 |   SwigValueWrapper& operator=(const T& t) { SwigSmartPointer tmp(new T(t)); pointer = tmp; return *this; }
```
See also https://open.cdash.org/viewBuildError.php?buildid=9511958

So I'm wondering, does SWIG support passing an `std::unique_ptr` parameter "by-value"? Or otherwise, should the `unique_ptr` be passed by r-value reference (`&&`), in order to make SWIG happy?

I'm asking, because the example by @christophe-calmejane at https://github.com/swig/swig/issues/2650 also uses `&&`.

**Comment by wsfulton** [comment]
The commits leading up to 3f1e40d2f4b89d04bddf430a7fa735f86b88cc3f included move semantics support for inputs (not outputs). Passing by value is also supported. I suggest you read the updated docs [unique_ptr smart pointer](https://htmlpreview.github.io/?https://github.com/swig/swig/blob/master/Doc/Manual/Library.html#Library_std_unique_ptr) and provide a cut down testcase if not working for you.

**Reply by N-Dekker** [reply]
Thank you for your reply @wsfulton I'm glad to hear that SWIG _does_ support pass by value, for `unique_ptr`. So it's still unclear to me why that particular C++ code that I proposed for ITK did not compile. My workaround to add `&&` did solve the problem: commit https://github.com/InsightSoftwareConsortium/ITK/pull/4533/commits/d39d95979293d683b9844b8f7dbba98418098c03

I'll have another look!

----

Just for the record: While it appears that there is no consensus on how to preferably pass a `unique_ptr`, the _C++ Core Guidelines_ certainly prefer pass by value over pass by `&&`, for `unique_ptr`: [Prefer simple and conventional ways of passing information](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-conventional)

**Comment by N-Dekker** [comment]
Update: I reproduced the compile errors, using swigwin-4.2.1, and the following "i" file:

```
/* File : my_unique_struct.i */
%module my_unique_struct
%include <std_unique_ptr.i>
%inline %{
#include <memory>
struct MyStruct {};
inline void take_my_unique_struct(std::unique_ptr<MyStruct> my_unique_struct)
{
}
%}
```
For which `swig -python -c++ my_unique_struct.i` generates:

```cpp
SWIGINTERN PyObject *_wrap_take_my_unique_struct(PyObject *self, PyObject *args) {
  PyObject *resultobj = 0;
  SwigValueWrapper< std::unique_ptr< MyStruct > > arg1 ;
  void *argp1 ;
  int res1 = 0 ;
  PyObject *swig_obj[1] ;
  
  (void)self;
  if (!args) SWIG_fail;
  swig_obj[0] = args;
  {
    res1 = SWIG_ConvertPtr(swig_obj[0], &argp1, SWIGTYPE_p_std__unique_ptrT_MyStruct_t,  0  | 0);
    if (!SWIG_IsOK(res1)) {
      SWIG_exception_fail(SWIG_ArgError(res1), "in method '" "take_my_unique_struct" "', argument " "1"" of type '" "std::unique_ptr< MyStruct >""'"); 
    }  
    if (!argp1) {
      SWIG_exception_fail(SWIG_ValueError, "invalid null reference " "in method '" "take_my_unique_struct" "', argument " "1"" of type '" "std::unique_ptr< MyStruct >""'");
    } else {
      std::unique_ptr< MyStruct > * temp = reinterpret_cast< std::unique_ptr< MyStruct > * >(argp1);
      arg1 = *temp;
      if (SWIG_IsNewObj(res1)) delete temp;
    }
  }
  take_my_unique_struct(SWIG_STD_MOVE(arg1));
  resultobj = SWIG_Py_Void();
  return resultobj;
fail:
  return NULL;
}
``` 

For which Visual C++ produces:
> my_unique_struct_wrap.cxx(3200,71): error C2280: 'std::unique_ptr<MyStruct,std::default_delete<MyStruct>>::unique_ptr(const std::unique_ptr<MyStruct,std::default_delete<MyStruct>> &)': attempting to reference a deleted function

As triggered by the assignment, `arg1 = *temp`.

The problem is that I did not explicitly wrap `std::unique_ptr<MyStruct>`, right? I should have added `%unique_ptr(MyStruct)` to the "i" file, right?

Does that mean that it's easier to just pass `unique_ptr` by rvalue-reference (`&&`), instead of by value? Should `&&` then be the preferred way to pass `unique_ptr` parameters, for SWIG users?

**Comment by christophe-calmejane** [comment]
> inline void take_my_unique_struct(std::unique_ptr<MyStruct> my_unique_struct)

This is not valid C++ code, so it makes sense that SWIG is not able to wrap it.

**Reply by N-Dekker** [reply]
Thanks @christophe-calmejane but why do you think `take_my_unique_struct(std::unique_ptr<MyStruct>)` is not valid?

**Reply by christophe-calmejane** [reply]
Because a std::unique_ptr is not copyable (deleted copy_ctor and copy_operator). That is why you get the compilation error.

**Reply by N-Dekker** [reply]
@christophe-calmejane When we ignore SWIG for a moment, the following does compile at https://godbolt.org/z/5Tc1Er1Gs

```cpp
#include <memory>
struct MyStruct {};
inline void take_my_unique_struct(std::unique_ptr<MyStruct> my_unique_struct)
{
}

int main()
{
    take_my_unique_struct(std::unique_ptr<MyStruct>());
    take_my_unique_struct(std::make_unique<MyStruct>());
}
```

So even while std::unique_ptr is not copyable, it may still be passed by value!  I _think_ that SWIG just produced those errors, because I did not add `%unique_ptr(MyStruct)` to my "i" file. But then, is it really necessary to add such a `%unique_ptr(...)` statement for each and every `std::unique_ptr<T>` that appears in the interface?

**Reply by christophe-calmejane** [reply]
Your code compiled because it is optimized out by the compiler (your method is empty). There is no discussion that a unique pointer is not copyable, meaning it cannot be passed by value. SWIG cannot do anything about that.

**Reply by N-Dekker** [reply]
[Section F.18](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f18-for-will-move-from-parameters-pass-by-x-and-stdmove-the-parameter) of the C++ Core Guidelines by Bjarne Stroustrup and Herb Sutter says:

> Unique owner types that are move-only and cheap-to-move, such as `unique_ptr`, can also be passed by value

And actually they recommend passing `unique_ptr` by value! Note that there has been some discussion, for example, Scott Meyers, July 24, 2014: [Should move-only types ever be passed by value?](http://scottmeyers.blogspot.com/2014/07/should-move-only-types-ever-be-passed.html)

**Reply by wsfulton** [reply]
@N-Dekker your simple C++ code is valid C++ and compiles because you pass a temporary to `take_my_unique_struct`. The compiler will then use the move semantics by default for which there is a move constructor available. It will thus compile.

Regarding SWIG, if you don't provide the definition of `std::unique_ptr` via std_unique_ptr.i, then SWIG assumes a copy constructor is available, which is incorrect, and so the code does not compile. Hence you must use `%unique_ptr(MyStruct)` in order to get this to work (which instantiates the `unique_ptr<MyStruct>` template for the SWIG type system and provides the appropriate typemaps. It also provides the machinery to call the move constructor instead of a copy constructor, that's all documented.

I would favour passing by value instead of rvalue reference as passing `std::unique_ptr` is widely understood that ownership of the underlying pointer is being transferred into the function and is a primary intended use case for this smart pointer. Without `%unique_ptr`, the default rvalue reference typemaps are used, but they don't understand this smart pointer and without the template instantiation it is being passed as an opaque type.

**Reply by christophe-calmejane** [reply]
Sorry for the confusion when I meant the code is not valid, it just not the usual use case. As pointed out by William it does compile _in your specific case_ because you pass a temporary to the method.
Regarding the core guidelines while I do agree that using by-value syntax does work for move-only types and the guidelines actually say "_it can be passed_" (not that it _should_), I prefer declaring methods with r-value reference instead so there is no ambiguity (but that is a personal preference and out of the scope of your question) which actually helps the caller to better diagnose compilation errors (it will fail right away if you try to pass an l-value, whilst it will fail in a chain of _deleted methods_ errors in a huge amount of error lines).

Regarding your question, I don't know if that helps but I usually define this (I've not done any SWIG for a long time, not an expert):

```
%typemap(out, optimal="1") SWIGTYPE
%{
  $result = new $1_ltype($1);
%}
```

**Comment by N-Dekker** [comment]
@wsfulton In the case I described at https://github.com/swig/swig/discussions/2854#discussioncomment-8997285, SWIG generated:

```cpp
std::unique_ptr< MyStruct > * temp = reinterpret_cast< std::unique_ptr< MyStruct > * >(argp1);
arg1 = *temp;
```
with local variable `arg1` being declared as `SwigValueWrapper< std::unique_ptr< MyStruct > >`

Shouldn't that assignment (`arg1 = *temp`) be a move-assignment instead? As in:

```cpp
arg1 = std::move(*temp);
```

**Reply by wsfulton** [reply]
I think my post crossed yours. Hopefully that clears this question up.

**Reply by N-Dekker** [reply]
Thanks @wsfulton It's _almost_ clear to me now 😃  Just double-checking... suppose SWIG would always generate a move-assignment, `arg1 = std::move(*temp)`, instead of `arg1 = *temp`, in a context like this. (`arg1` being a local variable of a `SwigValueWrapper<T>` type.) Wouldn't that be OK anyway?

But then again, you say that without `%unique_ptr(MyStruct)`, the generated type would be an "opaque type", so does that mean that it would not be useful anyway, _even_ _if_ my example code would compile without `%unique_ptr(MyStruct)`?

In [my ITK pull request](https://github.com/InsightSoftwareConsortium/ITK/pull/4533), I could silence the compile errors by passing the `std::unique_ptr` by `&&`. Do I understand correctly that that just _hides_ the problem?

**Reply by wsfulton** [reply, marked as answer]
SWIG tries very hard to generate code that compiles when type information is missing. This is different to a C++ compiler which absolutely requires full type information according to C++ rules. While SWIG will generate code that usually, but not always, compiles with missing type information, it is definitely not going to always be useful and you are nearly always better off providing the full type information. I suggest you look at the generated code with and without `%unique_ptr` to decide if omitting the type information meets your needs.

> suppose SWIG would always generate a move-assignment, arg1 = std::move(*temp), instead of arg1 = *temp

This may move the object's ownership into the C++ layer. Nearly always the wrong thing to do as the object is mostly needed again in the target language and so not a good default. The SWIGTYPE MOVE typemaps are available to force a move, see https://swig.org/Doc4.2/CPlusPlus11.html#CPlusPlus11_rvalue_reference_and_move_semantics, which also discusses move semantics in great detail for SWIG users.

**Comment by N-Dekker** [comment]
Thansk @wsfulton

For the record, this is the code generated by `swig -python -c++ my_unique_struct.i` when I add the missing `%unique_ptr(MyStruct)` to "my_unique_struct.i":

```cpp
SWIGINTERN PyObject *_wrap_take_my_unique_struct(PyObject *self, PyObject *args) {
  PyObject *resultobj = 0;
  std::unique_ptr< MyStruct > arg1 ;
  void *argp1 = 0 ;
  int res1 = 0 ;
  PyObject *swig_obj[1] ;
  
  (void)self;
  if (!args) SWIG_fail;
  swig_obj[0] = args;
  res1 = SWIG_ConvertPtr(swig_obj[0], &argp1, SWIGTYPE_p_MyStruct, SWIG_POINTER_RELEASE |  0 );
  if (!SWIG_IsOK(res1)) {
    if (res1 == SWIG_ERROR_RELEASE_NOT_OWNED) {
      SWIG_exception_fail(SWIG_ArgError(res1), "in method '" "take_my_unique_struct" "', cannot release ownership as memory is not owned for argument " "1"" of type '" "MyStruct *""'");
    } else {
      SWIG_exception_fail(SWIG_ArgError(res1), "in method '" "take_my_unique_struct" "', argument " "1"" of type '" "MyStruct *""'");
    }
  }
  (&arg1)->reset((MyStruct *)argp1);
  take_my_unique_struct(SWIG_STD_MOVE(arg1));
  resultobj = SWIG_Py_Void();
  return resultobj;
fail:
  return NULL;
}
```
And yes, then it does compile 👍

**Comment by N-Dekker** [comment]
> This may move the object's ownership into the C++ layer. Nearly always the wrong thing to do as the object is mostly needed again in the target language and so not a good default.

I was looking at the following generated code:
```cpp
      arg1 = *temp;
      if (SWIG_IsNewObj(res1)) delete temp;
```
_If_ the `delete temp` is executed, the object certainly isn't needed again, and the assignment might as well be a move-assignment. But it appears that that depends on `SWIG_IsNewObj(res1)`.  I don't know if it's worth rewriting those lines of code to something like:

```cpp
if (SWIG_IsNewObj(res1))
{
  // Move and delete
  arg1 = SWIG_STD_MOVE(*temp);
  delete temp;
}
else
  // Copy
  arg1 = *temp;
```

If copying `*temp` might be expensive, it might be worth considering... what do you think?

Anyway, with regard to my original question, it's clear to me now: for each `unique_ptr<T>` that is part of the API, a `%unique_ptr(T)` statement should be added, in order to properly swig the library.
